File: src/simulation/land.ts
import { islandRadius, landHeight, type Island } from '../maps/catalog';
import type { FleetActor } from './battle';
import type { Vec3 } from '../ships/blueprint';
import { clamp, wrapAngle, localToWorld } from './geometry';
import { motionVelocity, type HelmCommand } from './ship';
import { interpolate } from './hull';
import { damageHullContact, type HullImpact } from './contactDamage';

/** Sample keel stations against the same sloping seabed used by the map.
 * Draft and heading matter; the deep ocean remains navigable to submarines. */
export function resolveLandContact(actor: FleetActor, islands: readonly Island[], emit?: (impact: HullImpact) => void): void {
  const ship = actor.motion, hull = actor.definition.hull;
  for (const island of islands) {
    if (Math.hypot((ship.x - island.x) / (island.rx + hull.length), (ship.z - island.z) / (island.rz + hull.length)) > 1.3) continue;
    // Extend the authored shore slope into deep water. The visual terrain's
    // -45 m clipping floor is not a global submarine depth limit.
    const bottom = (x: number, z: number) => islandRadius(island, x, z) >= 1
      ? Math.max(-1000, (1 - islandRadius(island, x, z)) * 500) : landHeight([island], x, z);
    const points = hull.keelHeights.flatMap(([station, keel]) => [-.6, 0, .6].map(side =>
      localToWorld([side * interpolate(hull.halfBreadths, station), keel, hull.length / 2 - station], ship)));
    const hits = points.filter(point => bottom(point[0], point[2]) >= point[1]);
    if (!hits.length) continue;
    const position = hits[0];
    const dx = position[0] - island.x, dz = position[2] - island.z, distance = Math.hypot(dx, dz);
    const nx = distance > 1e-6 ? dx / distance : 1, nz = distance > 1e-6 ? dz / distance : 0;
    const velocity = motionVelocity(ship);
    const inward = -(velocity[0] * nx + velocity[2] * nz);
    if (inward > .75) {
      const damage = damageHullContact(actor, position, .5 * hull.massKg * inward ** 2);
      if (damage > 0) emit?.({ actor, position, damage, kind: 'grounding' });
    }
    if (inward > 0) { ship.speed = 0; ship.swaySpeed = 0; }
    // Move to the closest safe position along the outward direction. The
    // search includes keel depth, so shallow-draft ships approach more closely.
    const clear = (offset: number) => points.every(p => bottom(p[0] + nx * offset, p[2] + nz * offset) < p[1] - .02);
    let low = 0, high = Math.max(island.rx, island.rz) * 3 + hull.length;
    for (let i = 0; i < 24; i++) { const mid = (low + high) / 2; if (clear(mid)) high = mid; else low = mid; }
    ship.x += nx * high; ship.z += nz * high;
  }
}

/** Blend an outward course into the bot's intended route before it reaches shore. */
export function avoidLand(actor: FleetActor, command: HelmCommand, islands: readonly Island[]): HelmCommand {
  const ship = actor.motion;
  for (const island of islands) {
    const look = Math.max(650, Math.abs(ship.speed) * 50);
    const x = ship.x + Math.sin(ship.heading) * look, z = ship.z - Math.cos(ship.heading) * look;
    if (islandRadius(island, x, z) > 1.3 && islandRadius(island, ship.x, ship.z) > 1.2) continue;
    const away = Math.atan2(ship.x - island.x, island.z - ship.z);
    const rudder = clamp(wrapAngle(away - ship.heading) * 2, -1, 1);
    return { ...command, throttle: .4, rudder: Math.abs(rudder) < .05 ? 1 : rudder };
  }
  return command;
}

/** Swept terrain query, refined to the first intersection. Shared by shells and torpedoes. */
export function firstLandHit(islands: readonly Island[], from: Vec3, to: Vec3): { t: number; point: Vec3 } | undefined {
  if (!islands.length) return;
  const near = islands.filter(island => Math.min(from[0], to[0]) <= island.x + island.rx * 1.2 && Math.max(from[0], to[0]) >= island.x - island.rx * 1.2
    && Math.min(from[2], to[2]) <= island.z + island.rz * 1.2 && Math.max(from[2], to[2]) >= island.z - island.rz * 1.2);
  if (!near.length) return;
  const point = (t: number): Vec3 => [from[0] + (to[0] - from[0]) * t, from[1] + (to[1] - from[1]) * t, from[2] + (to[2] - from[2]) * t];
  const solid = (p: Vec3) => p[1] <= landHeight(near, p[0], p[2]);
  const steps = Math.max(1, Math.ceil(Math.hypot(to[0] - from[0], to[2] - from[2]) / 20));
  if (solid(from)) return { t: 0, point: [...from] };
  for (let i = 1; i <= steps; i++) if (solid(point(i / steps))) {
    let a = (i - 1) / steps, b = i / steps;
    for (let j = 0; j < 16; j++) { const mid = (a + b) / 2; if (solid(point(mid))) b = mid; else a = mid; }
    return { t: b, point: point(b) };
  }
}

File: crates/naval-sim/src/land.rs
//! Grounding: hulls against the map's terrain. A hull point at or below the ground
//! (`Terrain::height`, the sea floor included) puts the ship aground; the contact
//! point's escape direction is the push-out normal, the closing speed along it
//! scores contact damage, and the ship is moved out along it until every point
//! clears.
use crate::{
    collisions::{HullImpact, damage_hull_contact},
    definition::{Hull, Vec3},
    hull::interpolate,
    terrain::Terrain,
    vessel::Vessel,
};

/// Ship-local points that can touch the ground: three across each keel station of
/// an authored hull, or every vertex of a downward-facing constructed surface.
pub fn ground_points(hull: &Hull) -> Vec<Vec3> {
    if let Some(v) = &hull.volume {
        return v
            .surfaces
            .iter()
            .filter(|s| s.normal[1] < 0.)
            .flat_map(|s| s.vertices.iter().copied())
            .collect();
    }
    hull.keel_heights
        .iter()
        .flat_map(|[station, keel]| {
            [-0.6, 0.0, 0.6].map(|side| {
                [
                    side * interpolate(&hull.half_breadths, *station),
                    *keel,
                    hull.length / 2.0 - station,
                ]
            })
        })
        .collect()
}

/// Largest push-out tried, metres: only a hull placed deep inland needs more.
const MAX_PUSH_M: f64 = 4096.0;

pub fn resolve_land_contact(actor: &mut Vessel, terrain: &Terrain) -> Vec<HullImpact> {
    if !terrain.has_land() {
        return vec![];
    }
    let compiled = actor.compiled.clone();
    let hull = &compiled.definition.hull;
    let basis = actor.motion.basis();
    // The hull's transformed bounding box: when its lowest corner clears the
    // highest ground under its footprint, no point can be aground.
    let [low, high] = compiled.ground_box;
    let (mut floor, mut x0, mut z0, mut x1, mut z1) = (
        f64::INFINITY,
        f64::INFINITY,
        f64::INFINITY,
        f64::NEG_INFINITY,
        f64::NEG_INFINITY,
    );
    for corner in 0..8 {
        let local = std::array::from_fn(|i| {
            if corner >> i & 1 == 0 {
                low[i]
            } else {
                high[i]
            }
        });
        let p = basis.local_to_world(local);
        floor = floor.min(p[1]);
        (x0, z0, x1, z1) = (x0.min(p[0]), z0.min(p[2]), x1.max(p[0]), z1.max(p[2]));
    }
    if terrain.highest(x0, z0, x1, z1) < floor {
        return vec![];
    }
    let points: Vec<Vec3> = compiled
        .ground_points
        .iter()
        .map(|p| basis.local_to_world(*p))
        .collect();
    let bottom = |x: f64, z: f64| terrain.height(x, z);
    let Some(position) = points.iter().find(|p| bottom(p[0], p[2]) >= p[1]).copied() else {
        return vec![];
    };
    // Away from the coast at the contact. A contact already up on the land has
    // its own nearest land sample underfoot, so a hull still afloat amidships
    // takes the way out from its centre instead.
    let from = if bottom(position[0], position[2]) > 0.0
        && bottom(actor.motion.x, actor.motion.z) <= 0.0
    {
        [actor.motion.x, actor.motion.z]
    } else {
        [position[0], position[2]]
    };
    let [nx, nz] = terrain.escape(from[0], from[1]);
    let mut impacts = vec![];
    // Contact energy and the body-speed round trip below use physical motion.
    let ship = &actor.motion;
    let velocity = ship.physical_velocity();
    let inward = -(velocity[0] * nx + velocity[2] * nz);
    if inward > 0.75 {
        let damage = damage_hull_contact(
            actor,
            &compiled.definition,
            position,
            0.5 * hull.mass_kg * inward.powi(2),
        );
        if damage > 0.0 {
            impacts.push(HullImpact {
                actor_id: actor.motion.id.clone(),
                other_id: None,
                position,
                damage,
                kind: "grounding".into(),
            });
        }
    }
    let ship = &mut actor.motion;
    if inward > 0.0 {
        let vx = velocity[0] + nx * inward - ship.drift_x;
        let vz = velocity[2] + nz * inward - ship.drift_z;
        ship.speed = ship.heading.sin() * vx - ship.heading.cos() * vz;
        ship.sway_speed = ship.heading.cos() * vx + ship.heading.sin() * vz;
        if ship.speed.abs() < 1e-9 {
            ship.speed = 0.0;
        }
        if ship.sway_speed.abs() < 1e-9 {
            ship.sway_speed = 0.0;
        }
    }
    let clear = |offset: f64| {
        points
            .iter()
            .all(|p| bottom(p[0] + nx * offset, p[2] + nz * offset) < p[1] - 0.02)
    };
    // Bracket the push-out, then bisect it as finely as before.
    let mut high = 8.0;
    while high < MAX_PUSH_M && !clear(high) {
        high *= 2.0;
    }
    let mut low = 0.0;
    for _ in 0..24 {
        let mid = (low + high) / 2.0;
        if clear(mid) {
            high = mid;
        } else {
            low = mid;
        }
    }
    ship.x += nx * high;
    ship.z += nz * high;
    impacts
}
